include/pybind11/pybind11.h

```diff
@@ -967,11 +967,19 @@ class cpp_function : public function {
             return nullptr;
         }
         if (!result) {
-            std::string msg = "Unable to convert function return value to a "
-                              "Python type! The signature was\n\t";
-            msg += it->signature;
-            append_note_if_missing_header_is_suspected(msg);
-            PyErr_SetString(PyExc_TypeError, msg.c_str());
+            /* Allow the user to raise a Python exception directly using
+               PyErr_SetString() together with returning a null py::object, as
+               that may be significantly faster than throwing a C++ exception
+               in critical code paths. In that case we arrive here with
+               non-null PyErr_Occurred(), so keep that exception instead of
+               overwriting it with another. */
+            if (!PyErr_Occurred()) {
```

Comment:
I need to look at this more carefully and think about it.
First impression:
* we're building a surprising bypass here for the sake of performance.
* it employs a side effect: the wrapped function manipulating the global Python error state.
We're definitely setting ourself up for quicksand experiences: people thinking they are on firm ground, but there is a bug elsewhere that makes the Python error state invalid, but it only manifests itself through this bypass with an off-topic error that could take hours of debugging to understand. (I've been in similar situations many times.)
Is the benefit of this optimization so big that we're ok accepting that danger ("at scale" it is pretty much certain that people will run into it)?

Reply:
Maybe I should add: if performance *really* matters, why do you have this code path in your loop?
Of course it's nice if it's fast.
But there is an opportunity cost to optimizations like this PR: things get more complicated and fragile.
Now we're in the realm of philosophy:
* Keep pybind11 lean and straightforward so that it is as good as it can be at wrapping with ease.
* Tell users: if you have a performance critical loop, put your data into a C++ array (that you can easily wrap) and have your performance critical loop in C++, taking Python completely out of that loop, literally.
So there are clear "performance really doesn't matter" and "performance really does matter" regimes where everybody is in agreement and happy, but a gray area in the middle, where it's not quite important enough for users to bite the bullet and code up the C++ array they'd need.
What's better globally over a long period of time for all users in aggregate?
* Having N bypasses like this in pybind11 to get more of that gray area?
* A leaner and less fragile pybind11?

Reply:
Tagging #2760 and @jblespiau who is maybe more the the N bypasses side of the spectrum than I am.

Reply:
@rwgk This would also allow better error messages from type casters.

Reply:
Drive-by response:
> This would also allow better error messages from type casters.

That would change my cost-benefit equation (previously I was only seeing "performance" as the benefit).

A PR that generates better error messages, based on this PR, may be convincing.

Reply:
> Tagging #2760 and @jblespiau who is maybe more on the N bypasses side of the spectrum than I am.

The bypass I was suggesting in the past was mostly about not performing a *dynamic* look-up every time we do a C++ to Python or Python to C++ conversion. I don't really consider this a "bypass", but more of a cache (if we know the C++ Type "A" maps to a given Python type, no need to do the look-up in the hash-map every single time to retrieve the same object again and again. It's even more obvious when you convert an std::vector<A>, you will do as many look-ups as there are items in the vector. I was suggesting to add the keyword "static" on the look-up line, and you get a huge speed-up. The only possible issue is if people unregister and re-register new types on the same symbol, but I don't even think it's possible (you cannot unregister a symbol as far as I know).
Another difference with the current CL is that I was suggesting to improve the normal path, not an exceptional one.

Specifically for the current PR, I would probably need to think more, but I am thinking that:

- Exceptions are known to be slow in Python. Exceptions should be exceptional, and not be the standard way to return something. 
- Most of the time, after an exception, we crash.
- Given they are expected to be exceptional, why do we want to optimize it in the first place?

So I am wondering (a) why try to optimize something that should not happen in normal circumstances, (b) and also why, according to the benchmark, it has any impact on the non-raising path : if no exception is raised, it shouldn't change the runtime, right?